In [1]:
# Importing essesntial Libraries
import pandas as pd
import numpy as np

In [2]:
data = pd.read_csv('train_cleaned.csv')
data.head()

Item_Weight  Item_Visibility  Item_MRP  Outlet_Establishment_Year  \
0         9.30         0.016047  249.8092                       1999   
1         5.92         0.019278   48.2692                       2009   
2        17.50         0.016760  141.6180                       1999   
3        19.20         0.000000  182.0950                       1998   
4         8.93         0.000000   53.8614                       1987   

   Item_Outlet_Sales  Item_Fat_Content_LF  Item_Fat_Content_Low Fat  \
0          3735.1380                    0                         1   
1           443.4228                    0                         0   
2          2097.2700                    0                         1   
3           732.3800                    0                         0   
4           994.7052                    0                         1   

   Item_Fat_Content_Regular  Item_Fat_Content_low fat  Item_Fat_Content_reg  \
0                         0                         0                     0   
1                         1                         0                     0   
2                         0                         0                     0   
3                         1                         0                     0   
4                         0                         0                     0   

   ...  Outlet_Size_High  Outlet_Size_Medium  Outlet_Size_Small  \
0  ...                 0                   1                  0   
1  ...                 0                   1                  0   
2  ...                 0                   1                  0   
3  ...                 0                   0                  0   
4  ...                 1                   0                  0   

   Outlet_Location_Type_Tier 1  Outlet_Location_Type_Tier 2  \
0                            1                            0   
1                            0                            0   
2                            1                            0   
3                            0                            0   
4                            0                            0   

   Outlet_Location_Type_Tier 3  Outlet_Type_Grocery Store  \
0                            0                          0   
1                            1                          0   
2                            0                          0   
3                            1                          1   
4                            1                          0   

   Outlet_Type_Supermarket Type1  Outlet_Type_Supermarket Type2  \
0                              1                              0   
1                              0                              1   
2                              1                              0   
3                              0                              0   
4                              1                              0   

   Outlet_Type_Supermarket Type3  
0                              0  
1                              0  
2                              0  
3                              0  
4                              0  

[5 rows x 46 columns]

In [3]:
data.shape

(8523, 46)

In [4]:
data.isnull().sum()

Item_Weight                        0
Item_Visibility                    0
Item_MRP                           0
Outlet_Establishment_Year          0
Item_Outlet_Sales                  0
Item_Fat_Content_LF                0
Item_Fat_Content_Low Fat           0
Item_Fat_Content_Regular           0
Item_Fat_Content_low fat           0
Item_Fat_Content_reg               0
Item_Type_Baking Goods             0
Item_Type_Breads                   0
Item_Type_Breakfast                0
Item_Type_Canned                   0
Item_Type_Dairy                    0
Item_Type_Frozen Foods             0
Item_Type_Fruits and Vegetables    0
Item_Type_Hard Drinks              0
Item_Type_Health and Hygiene       0
Item_Type_Household                0
Item_Type_Meat                     0
Item_Type_Others                   0
Item_Type_Seafood                  0
Item_Type_Snack Foods              0
Item_Type_Soft Drinks              0
Item_Type_Starchy Foods            0
Outlet_Identifier_OUT010           0
O

# Isolating Dependent and Independent Variables

In [5]:
x = data.drop('Item_Outlet_Sales', axis = 1)
y = data['Item_Outlet_Sales']
x.shape, y.shape

((8523, 45), (8523,))

# Dividing data into train-test

In [6]:
from sklearn.model_selection import train_test_split
trainx, testx, trainy, testy = train_test_split(x, y, random_state = 101)
trainx.shape, testx.shape, trainy.shape, testy.shape

((6392, 45), (2131, 45), (6392,), (2131,))

# Implementing Linear Regression

In [7]:
from sklearn.linear_model import LinearRegression
lreg= LinearRegression()
# Training Model
lreg.fit(trainx, trainy)
# Making Predictions
pred1 = lreg.predict(testx)
# Finding Score
m1score = lreg.score(testx, testy)
print(f'Prediction Score : {(m1score*100).round(2)}')

Prediction Score : 56.28


# Implementing KNN

In [8]:
from sklearn.neighbors import KNeighborsRegressor
knn = KNeighborsRegressor(n_neighbors=9)
# Training Model
knn.fit(trainx, trainy)
# Making Predictions
pred2 = knn.predict(testx)
# Finding Score
m2score = knn.score(testx, testy)
print(f'Prediction Score : {(m2score*100).round(2)}')

Prediction Score : 49.12


# Implementing Decision Tree Regressor

In [9]:
from sklearn.tree import DecisionTreeRegressor
dtr= DecisionTreeRegressor(max_depth=7)
# Training Model
dtr.fit(trainx, trainy)
# Making Predictions
pred3 = dtr.predict(testx)
# Finding Score
m3score = dtr.score(testx, testy)
print(f'Prediction Score : {(m3score*100).round(2)}')

Prediction Score : 57.47


# Averaging

In [10]:
from statistics import mean
final_preds = np.array([])
for i in range(len(testy)):
    final_preds = np.append(final_preds, mean([pred1[i], pred2[i], pred3[i]]))

In [11]:
from sklearn.metrics import r2_score
final_score = r2_score(testy, final_preds)
print(f'Prediction Score : {(final_score*100).round(2)}')

Prediction Score : 58.11


## Slight increase found
### Individiul Scores: [56.28, 49.12, 57.47]
### Final Score: 58.11

# Weighted Averaging

In [12]:
from statistics import mean
final_preds = np.array([])
for i in range(len(testy)):
    final_preds = np.append(final_preds, mean([pred1[i], pred1[i], pred2[i], pred3[i], pred3[i]]))

In [13]:
from sklearn.metrics import r2_score
final_score = r2_score(testy, final_preds)
print(f'Prediction Score : {(final_score*100).round(2)}')

Prediction Score : 58.69


## Slight increase found
### Averaging Scores: 58.11
### Weighted Averaging Score: 58.69

# Ranked Averaging

In [16]:
scores = [m1score , m2score, m3score]
ranks = pd.DataFrame({'Scores' : scores}) 
ranks.head()

Scores
0  0.562822
1  0.491197
2  0.574706

In [23]:
ranks = ranks.sort_values('Scores')
ranks.head()

Scores
1  0.491197
0  0.562822
2  0.574706

In [26]:
ranks['rank'] = [x for x in range(1,ranks.shape[0]+1) ]
ranks.head()

Scores  rank
1  0.491197     1
0  0.562822     2
2  0.574706     3

In [27]:
ranks['weights'] = ranks['rank']/ranks['rank'].sum()
ranks.head()

Scores  rank   weights
1  0.491197     1  0.166667
0  0.562822     2  0.333333
2  0.574706     3  0.500000

In [44]:
from statistics import mean
final_preds = np.array([])
for i in range(len(testy)):
    final_preds = np.append(final_preds, np.sum([pred1[i]*ranks['weights'][0], pred2[i]*ranks['weights'][1], ranks['weights'][2]*pred3[i]]))

In [45]:
from sklearn.metrics import r2_score
final_score = r2_score(testy, final_preds)
print(f'Prediction Score : {(final_score*100).round(2)}')

Prediction Score : 58.81


## Slight increase found
### Averaging Scores: 58.11
### Weighted Averaging Score: 58.69
### Ranked Averaging Score: 58.81